keras/utils/hdf5_utils.py

```diff
@@ -0,0 +1,197 @@
+    h5py = None
+
+
+if sys.version_info[0] == 3:
+    import pickle
+else:
+    import cPickle as pickle
+
+
+# FLAGS
+_is_boxed = '_is_dict_boxed'
+
+
+class H5Dict(object):
+    """ A dict-like wrapper around h5py groups (or dicts).
+
+    This allows us to have a single serialization logic
+    for both pickling and saving to disk.
+
+    Note: This is not intended to be a generic wrapper.
+    There are lot of edge cases which have been hardcoded,
+    and makes sense only in the context of model serialization/
+    deserialization.
+    """
+
+    def __init__(self, path, mode='a'):
+        if isinstance(path, h5py.Group):
+            self.data = path
+            self._is_file = False
+        elif type(path) is str:
+            self.data = h5py.File(path,)
+            self._is_file = True
+        elif type(path) is dict:
+            self.data = path
+            self._is_file = False
+            self.data[_is_boxed] = True
+        else:
+            raise Exception('Required Group, str or dict.'
+                            ' Received {}.'.format(type(path)))
+        self.read_only = mode == 'r'
+
+    def __setitem__(self, attr, val):
+        if self.read_only:
+            raise Exception('Can not set item in read only mode.')
+        is_np = type(val).__module__ == np.__name__
+        if type(self.data) is dict:
+            if type(attr) is bytes:
+                attr = attr.decode('utf-8')
+            if is_np:
+                self.data[attr] = pickle.dumps(val)
+                # We have to remeber to unpickle in __getitem__
+                self.data['_{}_pickled'.format(attr)] = True
+            else:
+                self.data[attr] = val
+            return
+        if attr in self:
+            raise Exception('Can not set attribute.'
+                            ' Group with name {} exists.'.format(attr))
+        if is_np:
+            dataset = self.data.create_dataset(attr, val.shape, dtype=val.dtype)
+            if not val.shape:
+                # scalar
+                dataset[()] = val
+            else:
+                dataset[:] = val
+        if isinstance(val, list):
+            # Check that no item in `data` is larger than `HDF5_OBJECT_HEADER_LIMIT`
+            # because in that case even chunking the array would not make the saving
+            # possible.
+            bad_attributes = [x for x in val if len(x) > HDF5_OBJECT_HEADER_LIMIT]
+
+            # Expecting this to never be true.
+            if len(bad_attributes) > 0:
+                raise RuntimeError('The following attributes cannot be saved to HDF5 '
+                                   'file because they are larger than %d bytes: %s'
+                                   % (HDF5_OBJECT_HEADER_LIMIT,
+                                      ', '.join([x for x in bad_attributes])))
+
+            data_npy = np.asarray(val)
+
+            num_chunks = 1
+            chunked_data = np.array_split(data_npy, num_chunks)
```

Comment:
Can we have a better initial `num_chunks`?
For example, `math.ceil(data_npy.nbytes / HDF5_OBJECT_HEADER_LIMIT)`